tensorflow/compiler/tf2tensorrt/utils/funcdef_to_graphdef.cc

```diff
@@ -0,0 +1,142 @@
+  // This is the same function as in function.cc. However, it uses the
+  // name mapping above, which retains IO prefixes (prefixes.kInputPHName etc)
+  gtl::InlinedVector<const Edge*, 4> inputs;
+  gdef->Clear();
+  *gdef->mutable_versions() = g->versions();
+
+  std::vector<Node*> start_nodes;
+  for (Node* n : g->nodes()) {
+    if (n->out_edges().empty()) {
+      start_nodes.push_back(n);
+    }
+  }
+
+  ReverseDFSFrom(*g, start_nodes, nullptr, [gdef, &inputs](Node* n) {
+    if (!n->IsOp()) return;
+    NodeDef* ndef = gdef->add_node();
+    ndef->set_name(AppendIdToNodeName(n));
+    ndef->set_op(n->type_string());
+    for (const auto& attr : n->attrs()) {
+      (*ndef->mutable_attr())[attr.first] = attr.second;
+    }
+
+    if (!n->assigned_device_name().empty()) {
+      ndef->set_device(n->assigned_device_name());
+    } else {
+      ndef->set_device(n->requested_device());
+    }
+
+    inputs.clear();
+    inputs.resize(n->num_inputs());
+    for (const Edge* e : n->in_edges()) {
+      if (e->IsControlEdge()) {
+        inputs.push_back(e);
+      } else {
+        if (inputs[e->dst_input()] == nullptr) {
+          inputs[e->dst_input()] = e;
+        } else {
+          LOG(WARNING) << "Malformed graph node. multiple input edges: "
+                       << n->DebugString();
+        }
+      }
+    }
+    // node->name() is merely NodeDef::name, which are not guaranteed
+    // to be unique and stable after optimization rewrites. Therefore,
+    // we use "n<node id> or <io prefix><node_id>" instead.
+    for (const Edge* e : inputs) {
+      if (e == nullptr) {
+        ndef->add_input("unknown");
+        continue;
+      }
+      const string srcname = AppendIdToNodeName(e->src());
+      if (!e->src()->IsOp()) {
+      } else if (e->IsControlEdge()) {
+        ndef->add_input(strings::StrCat("^", srcname));
+      } else if (e->src_output() == 0) {
+        ndef->add_input(srcname);
+      } else {
+        ndef->add_input(strings::StrCat(srcname, ":", e->src_output()));
+      }
+    }
+  });
+}
+
+Status FunctionDefToGraphDef(FunctionLibraryRuntime::Handle handle,
+                             FunctionLibraryRuntime* flib_runtime,
+                             GraphDef* graph_def,
+                             std::vector<int>* input_node_ids,
+                             std::vector<int>* output_node_ids) {
+  const FunctionLibraryDefinition* flib_def =
+      flib_runtime->GetFunctionLibraryDefinition();
+  const FunctionBody* fbody;
+  fbody = flib_runtime->GetFunctionBody(handle);
+  if (!fbody) {
+    return errors::Internal(
+        "Function body is null when converting from FuncDef to GraphDef.");
+  }
+  std::unique_ptr<Graph> graph(new Graph(flib_def));
+
+  CopyGraph(*fbody->graph, graph.get());
+
+  for (Node* n : graph->nodes()) {
+    auto id = n->id();
+    if (n->IsArg()) {
+      VLOG(2) << "Arg Node id used for unique naming is " << id;
+      input_node_ids->push_back(id);
```

Comment:
`graph` is a temporary object, why do we use the node id as the id of the input/output nodes?